### PR TITLE
Fix broken caching handling of job renames by purging cache entry [JENKINS-38159]

### PR DESCRIPTION
**Pull request by svanoort**
[JENKINS-38159](https://issues.jenkins-ci.org/browse/JENKINS-38159) - log URLs would have to be rewritten (as well as all HAL links in JSON), so it's better to just purge the cache entry and rebuild when needed.

### CODE REVIEW

**Comment by svanoort** [comment]
@reviewbybees

File: rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java [hunk 1]
```diff
@@ -267,36 +267,33 @@ public static WorkflowRun getWorkflowRunForExecution(@CheckForNull FlowExecution
...
-                    if (!r.isBuilding()) {
-                        String path = oldFullName+"#"+r.getId();
-                        RunExt cachedRun = rc.getIfPresent(path);
-                        if (cachedRun != null) {
-                            rc.invalidate(path);
-                            rc.put(newFullName+"#"+r.getId(), cachedRun);
-                        }
-                    }
+            ConcurrentMap<String, RunExt> runMap = rc.asMap();
+            for (String cacheRunId : runMap.keySet()) {  // Put the Concurrent in ConcurrentMap to work for us
```
Comment (jglick) [inline rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java]: Maybe better style to use `Iterator.remove`.

File: rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java [hunk 2]
```diff
@@ -267,36 +267,33 @@ public static WorkflowRun getWorkflowRunForExecution(@CheckForNull FlowExecution
...
+                // Null-check may not be needed, but just in case of mutation by another thread
```
Comment (jglick) [inline rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java]: Would that really result in null entries??

Comment (svanoort) [inline rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java]: @jglick  I hope not, but am feeling a bit cautious - plus it's easier to put the check than to prove it safe.

**Pull request merged**